cirq-ft/cirq_ft/infra/t_complexity_protocol.py

```diff
@@ -0,0 +1,193 @@
+# Copyright 2023 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Callable, Hashable, Iterable, Optional, Union, overload
+from typing_extensions import Literal
+
+import attr
+import cachetools
+import cirq
+from cirq_ft.infra.decompose_protocol import _decompose_once_considering_known_decomposition
+from typing_extensions import Protocol
+
+_T_GATESET = cirq.Gateset(cirq.T, cirq.T**-1, unroll_circuit_op=False)
+
+
+@attr.frozen
+class TComplexity:
+    t: int = 0
+    clifford: int = 0
+    rotations: int = 0
+
+    def __add__(self, other: 'TComplexity') -> 'TComplexity':
+        return TComplexity(
+            self.t + other.t, self.clifford + other.clifford, self.rotations + other.rotations
+        )
+
+    def __mul__(self, other: int) -> 'TComplexity':
+        return TComplexity(self.t * other, self.clifford * other, self.rotations * other)
+
+    def __rmul__(self, other: int) -> 'TComplexity':
+        return self.__mul__(other)
+
+    def __str__(self) -> str:
+        return (
+            f'T-count:   {self.t:g}\n'
+            f'Rotations: {self.rotations:g}\n'
+            f'Cliffords: {self.clifford:g}\n'
+        )
+
+
+class SupportsTComplexity(Protocol):
```

Comment:
Should this be in a protocol directory to stay consistent with cirq?

Reply:
Going forward, I expect a lot more algorithm specific directories than just an `ops` (like in Cirq). I'd like to punt the directory reorganization to a later time once the library grows to a bigger scale. For now, `infra/` and `algos/` seem to be enough for top-level directories.